pandas/core/indexes/base.py

```diff
@@ -2858,13 +2858,17 @@ def _intersection(self, other, sort=False):
             indexer = algos.unique1d(Index(rvals).get_indexer_non_unique(lvals)[0])
             indexer = indexer[indexer != -1]
 
-        result = other.take(indexer).unique()._values
+        result = other.take(indexer)
+        if not result.is_unique:
+            result = result.unique()._values
+        else:
+            result = result._values
 
         if sort is None:
             result = algos.safe_sort(result)
 
         # Intersection has to be unique
-        assert algos.unique(result).shape == result.shape
+        assert Index(result).is_unique
```

Comment:
this is more performant than algos.unique version?  The Index constructor is non-trivial overhead

Reply:
Asvs for index_object did not show a slowdown, but could change it back the old version is preferred